Mathlib/CategoryTheory/Sites/ConstantSheaf.lean

```diff
@@ -45,22 +66,171 @@ sheafification of the constant presheaf.
 noncomputable def constantSheaf : D ⥤ Sheaf J D := Functor.const Cᵒᵖ ⋙ (presheafToSheaf J D)
 
 /-- The constant sheaf functor is left adjoint to evaluation at a terminal object. -/
+@[simps! counit_app]
 noncomputable def constantSheafAdj {T : C} (hT : IsTerminal T) :
     constantSheaf J D ⊣ (sheafSections J D).obj (op T) :=
   (constantPresheafAdj D hT).comp (sheafificationAdjunction J D)
 
-lemma constantSheafAdj_counit_app {T : C} (hT : IsTerminal T) (F : Sheaf J D) :
-    (constantSheafAdj J D hT).counit.app F =
-      (presheafToSheaf J D).map ((constantPresheafAdj D hT).counit.app F.val) ≫
-        (sheafificationAdjunction J D).counit.app F := by
+variable {D}
+
+namespace Sheaf
+
+/--
+A sheaf is constant if it is in the essential image of the constant sheaf functor.
+-/
+class IsConstant (F : Sheaf J D) : Prop where
+  mem_essImage : F ∈ (constantSheaf J D).essImage
+
+lemma mem_essImage_of_isConstant (F : Sheaf J D) [IsConstant J F] :
+    F ∈ (constantSheaf J D).essImage :=
+  IsConstant.mem_essImage
+
+lemma isConstant_congr {F G : Sheaf J D} (i : F ≅ G) [IsConstant J F] : IsConstant J G where
+  mem_essImage := essImage.ofIso i F.mem_essImage_of_isConstant
+
+lemma isConstant_of_iso {F : Sheaf J D} {X : D} (i : F ≅ (constantSheaf J D).obj X) :
+    IsConstant J F := ⟨_, ⟨i.symm⟩⟩
+
+lemma isConstant_iff_mem_essImage {L : D ⥤ Sheaf J D} {T : C} (hT : IsTerminal T)
+    (adj : L ⊣ (sheafSections J D).obj ⟨T⟩)
+    (F : Sheaf J D) : IsConstant J F ↔ F ∈ L.essImage := by
+  rw [essImage_eq_of_natIso (adj.leftAdjointUniq (constantSheafAdj J D hT))]
+  exact ⟨fun ⟨h⟩ ↦ h, fun h ↦ ⟨h⟩⟩
+
+instance (F : Sheaf J D) [HasTerminal C]
+    [IsIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F] : IsConstant J F where
+  mem_essImage := ⟨_, ⟨asIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F⟩⟩
+
+instance [(constantSheaf J D).Faithful] [(constantSheaf J D).Full] (F : Sheaf J D)
+    [IsConstant J F] {T : C} (hT : IsTerminal T) :
+    IsIso ((constantSheafAdj J D hT).counit.app F) := by
+  rw [isIso_counit_app_iff_mem_essImage]
+  exact F.mem_essImage_of_isConstant
```

Comment:
This looks like a loop. I would suggest keeping as an instance only the version `... [IsConstant J F] ... : IsIso ...`.